File: CSharp/Boardgames/Shared/Networking/MessageSerializer.cs
using GameClient;

namespace Shared.Data;

// NOTE(rune): Tilsvarende klasse i Java/C# skal passe sammen.
public static class MessageSerializer {
    public static Message Deserialize(string s) {
        int idx = s.IndexOf('|');
        if (idx > 0 && idx < s.Length - 1) {
            string headString = s.Substring(0, idx);
            string bodyString = s.Substring(idx + 1);

            Head head = JsonUtil.FromJson<Head>(headString);
            string bodyTypeName = typeof(Messages).FullName + "+" + head.bodyType; // NOTE(rune): "+" betyder nested class.
            Type? bodyType = Type.GetType(bodyTypeName);
            if (bodyType == null) {
                throw new Exception("Invalid response body type from game server (body type was '" + head.bodyType + "').");
            }

            object body = JsonUtil.FromJson(bodyString, bodyType);
            return new Message(head, body);
        }
        else {
            throw new Exception("Invalid response from game server (could not find '|' marker).");
        }
    }

    public static string Serialize(Message m) {
        string headString = JsonUtil.ToJson(m.Head);
        string bodyString = JsonUtil.ToJson(m.Body);
        string s = headString + "|" + bodyString;
        return s;
    }
}

File: Java/boardgames/logic/src/main/java/boardgames/logic/networking/MessageSerializer.java
package boardgames.logic.networking;

import boardgames.logic.messages.Message;
import boardgames.logic.messages.Messages;
import boardgames.shared.util.JsonUtil;
import boardgames.shared.util.Log;
import com.google.gson.JsonSyntaxException;

// NOTE(rune): Tilsvarende klasse i Java/C# skal passe sammen.
public class MessageSerializer {
    public static Message deserialize(String s) {
        int idx = s.indexOf('|');
        if (idx > 0 && idx < s.length() - 1) {
            String headString = s.substring(0, idx);
            String bodyString = s.substring(idx + 1);
            try {
                Messages.Head head = JsonUtil.fromJson(headString, Messages.Head.class);
                String bodyTypeName = Messages.class.getName() + "$" + head.bodyType(); // NOTE(rune): "$" betyder nested class.
                Class<?> bodyType = Class.forName(bodyTypeName);
                Object body = JsonUtil.fromJson(bodyString, bodyType);
                return new Message(head, body);
            } catch (JsonSyntaxException | ClassNotFoundException e) {
                Log.error(e);
                return null;
            }
        } else {
            Log.error("Invalid message string received (could not find '|' marker). Message string was '" + s + "'");
            return null;
        }
    }

    public static String serialize(Message m) {
        String headString = JsonUtil.toJson(m.head());
        String bodyString = JsonUtil.toJson(m.body());
        String s = headString + "|" + bodyString;
        return s;
    }
}
